File: c/square-root/square_root.c
#include "square_root.h"
#include <stdint.h>
#include <math.h>

/* Using the Binary numeral system implementation from
 * https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Binary_numeral_system_(base_2)
 */

uint16_t square_root(uint32_t n) {
    // find b, the greatest power of 4 less than or equal to n
    uint32_t b = pow(4, floor(log(n) / log(4)));
    uint16_t x = 0;

    while (b != 0) {
        if (n >= x + b) {
            n = n - x - b;
            x = (x >> 1) + b;
        }
        else {
            x >>= 1;
        }
        b >>= 2;
    }

    return x;
}

File: dart/square-root/lib/square_root.dart
import 'dart:math';

/*
Using the Binary numeral system (base 2) from Wikipedia
https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Binary_numeral_system_%28base_2%29
*/

class SquareRoot {
  int squareRoot(int radicand) {
    // find `b`, the greatest power of 4 ≤ radicand
    var b = 4.pow( (log(radicand) / log(4)).floor() );
    var root = 0;
    while (b > 0) {
      if (radicand >= root + b) {
        radicand -= root + b;
        root = root ~/ 2 + b;
      }
      else {
        root ~/= 2;
      }
      b ~/= 4;
    }
    return root;
  }
}

extension IntMethods on int {
  int pow(int exponent) {
    var result = 1;
    for (var i = 1; i <= exponent; i++) result *= this;
    return result;
  }
}
